Bluetooth: host: Fix DB hash calculation

thoh-ot opened the pull request:
When hashing attributes for the UUID cases:
    BT_UUID_GATT_PRIMARY_VAL
    BT_UUID_GATT_SECONDARY_VAL
    BT_UUID_GATT_INCLUDE_VAL
    BT_UUID_GATT_CHRC_VAL
    BT_UUID_GATT_CEP_VAL

where (handle || UUID || value) is to be hashed, the worst case size of
attribute value is 19:

Characteristic declaration (BT_UUID_GATT_CHRC_VAL)
Attribute Value:
* Characteristic Properties: 1 octets
* Characteristic Value Handle: 2 octets
* Characteristic UUID: 2 or 16 octets

Update the data array to accomodate for the worst case size.

Fixes #39131

Signed-off-by: Thomas Ebert Hansen <thoh@oticon.com>

thoh-ot commented:
There's an issue in the EDTT GATT test where the contents of 'GATT Database.xml' no longer matches, as the hash calculation now computes a different number for the 'Database Hash' characteristic.

alwa-nordic reviewed (changes requested):
Changes have been agreed on in conversation.

carlescufi commented on subsys/bluetooth/host/gatt.c:
```diff
@@ -611,7 +611,7 @@ static uint8_t gen_hash_m(const struct bt_gatt_attr *attr, uint16_t handle,
 	struct gen_hash_state *state = user_data;
 	struct bt_uuid_16 *u16;
-	uint8_t data[16];
+	uint8_t data[19];
```
@thoh-ot can you please create the complete union and use that?

thoh-ot commented:
As mentioned in the issue some of the GATT EDTT tests are failing with the DB hash fix, because there's an generated XML file with expected GATT DB values in it, and it no longer matches. 

As the consultant who implemented that system is no longer in the company, I have done some archaeological digging in our version control system, and we found the following:

In a PTS release there is an Excel sheet, with multiple reference GATT databases, that are to be used during PTS testing, and also GATT EDTT testing as they mimic the PTS tests.

This Excel sheet  and been modified and Visual Basic for Application code added for generating a XML file for the GATT EDTT python test code, and .c & .h files for the corresponding GATT DB in the DUT bsim code.

I tried to get the VBA code running and running, but there seems to be some hardcoded paths and from the look of it, it would generate code for older types and apis.

Which PTS release the Excel sheet originally came from i don't know, but i can see that it doesn't match the newest PTS release, where things having added and moved around in reference databases to enforce certain behavior.

The issue is how do we wan't to approach this now, would try to revive the generation of XML file, c and h files or would we just hardcode the failing GATT DB hashes?

carlescufi commented:
> As mentioned in the issue some of the GATT EDTT tests are failing with the DB hash fix, because there's an generated XML file with expected GATT DB values in it, and it no longer matches.
> 
> As the consultant who implemented that system is no longer in the company, I have done some archaeological digging in our version control system, and we found the following:
> 
> In a PTS release there is an Excel sheet, with multiple reference GATT databases, that are to be used during PTS testing, and also GATT EDTT testing as they mimic the PTS tests.
> 
> This Excel sheet and been modified and Visual Basic for Application code added for generating a XML file for the GATT EDTT python test code, and .c & .h files for the corresponding GATT DB in the DUT bsim code.
> 
> I tried to get the VBA code running and running, but there seems to be some hardcoded paths and from the look of it, it would generate code for older types and apis.
> 
> Which PTS release the Excel sheet originally came from i don't know, but i can see that it doesn't match the newest PTS release, where things having added and moved around in reference databases to enforce certain behavior.
> 
> The issue is how do we wan't to approach this now, would try to revive the generation of XML file, c and h files or would we just hardcode the failing GATT DB hashes?

I am not sure how this is related to the code in this PR however, and to the discussion we are having regarding the structs and unions?

thoh-ot commented:
It's the reason for why the CI is failing, no matter if we use unions, structs or just increase the array length, the CI will never pass. 

Fixing the GATT DB hash calculation will render the generated files located in EDTT and tests/bluetooth/bsim_bt/edtt_ble_test_app/gatt_test_app/src/gatt invalid and thus the CI will fail.

carlescufi commented:
> It's the reason for why the CI is failing, no matter if we use unions, structs or just increase the array length, the CI will never pass.
> 
> Fixing the GATT DB hash calculation will render the generated files located in EDTT and tests/bluetooth/bsim_bt/edtt_ble_test_app/gatt_test_app/src/gatt invalid and thus the CI will fail.

Oh, I see. Thanks for the explanation @thoh-ot

carlescufi commented:
@thoh-ot even without fixing the .xml file and/or regenerating it, couldn't we merge the fix so at least we can test it with PTS and 128-bit UUIDs?

Thalley commented on subsys/bluetooth/host/gatt.c:
```diff
@@ -606,12 +606,48 @@ struct gen_hash_state {
+	 */
+	union {
+		uint16_t uuid16;
+		uint8_t  uuid128[16];
```
```suggestion
		uint8_t  uuid128[BT_UUID_SIZE_128];
```

Thalley commented on subsys/bluetooth/host/gatt.c:
```diff
@@ -606,12 +606,48 @@ struct gen_hash_state {
+		uint16_t value_handle;
+		union {
+			uint16_t uuid16;
+			uint8_t  uuid128[16];
```
```suggestion
			uint8_t  uuid128[BT_UUID_SIZE_128];
```

thoh-ot commented:
@carlescufi I have "shelfed" the generation tool, and instead updated the "Reference Database Hash" in the EDTT GATT test with the correct value and also added an implementation of the Bluetooth SIG GATT test "GATT/SR/GAS/BV-02-C" however the CI doesn't pull in the Python module update (Pycryptodome) needed for the EDTT GATT test change to fix the failing tests, any idea on how to proceed?

thoh-ot commented:
The Python module update requires an update to the CI docker image but to speed this pr along, i have decided to split the task in two and only focus on (1) in this PR.
(1) Fix Host GATT DB hash calculation and Update EDTT with corect DB hash
(2) Introduce new EDTT GATT test (with new Python module)

Thalley commented on subsys/bluetooth/host/gatt.c:
```diff
@@ -606,12 +606,48 @@ struct gen_hash_state {
+	struct {
+		uint16_t attribute_handle;
+		uint16_t end_group_handle;
+		uint16_t uuid16;
+
+	} __packed inc;
```
```suggestion
		uint16_t uuid16;
	} __packed inc;
```

thoh-ot replied:
Ack

Thalley reviewed (approved):
LGTM

thoh-ot commented:
Waiting for the EDTT PR https://github.com/EDTTool/EDTT/pull/50 to be merged and then mirrored over into https://github.com/zephyrproject-rtos/edtt, so i can update the edtt west sha to point to the correct sha.

alwa-nordic commented:
@thoh-ot The EDDT PR is merges now. Should we move this out of draft status?

thoh-ot committed:
tests: Bluetooth: bsim: Fix test names

The LL test for LLCP are using the old non-inclusive names.

Signed-off-by: Thomas Ebert Hansen <thoh@oticon.com>

thoh-ot committed:
tests: Bluetooth: bsim: Remove unknown tests

Remove unknown tests from the GATT test list.

Signed-off-by: Thomas Ebert Hansen <thoh@oticon.com>

thoh-ot committed:
manfiest: update edtt

Update edtt sha to 1fc2e6f7d2a01a01074456f2ef2ccf1f3df7fac1
>> Fix Database Hash value after GATT fix

Signed-off-by: Thomas Ebert Hansen <thoh@oticon.com>

wopu-ot reviewed (approved):
LGTM

thoh-ot commented:
@alwa-nordic have your 'request changes' been addressed?

alwa-nordic commented on subsys/bluetooth/host/gatt.c:
```diff
@@ -606,12 +606,47 @@ struct gen_hash_state {
+		union {
+			uint16_t uuid16;
+			uint8_t  uuid128[BT_UUID_SIZE_128];
+		};
```
Missing __packed here.

thoh-ot replied:
Fixed

thoh-ot committed:
Bluetooth: host: Fix DB hash calculation

When hashing attributes for the UUID cases:
    BT_UUID_GATT_PRIMARY_VAL
    BT_UUID_GATT_SECONDARY_VAL
    BT_UUID_GATT_INCLUDE_VAL
    BT_UUID_GATT_CHRC_VAL
    BT_UUID_GATT_CEP_VAL

where (handle || UUID || value) is to be hashed, the worst case size of
attribute value can be deduced from the associated union that describes
all the attribute values in the given cases.

Update the data array to accommodate for the worst case size.

Fixes #39131

Signed-off-by: Thomas Ebert Hansen <thoh@oticon.com>

thoh-ot commented:
> Changes have been agreed on in conversation.

@alwa-nordic Is everything good now?

jhedberg commented:
@alwa-nordic please take another look here. Your review is blocking merging.

carlescufi commented:
@thoh-ot thanks for following this up, this look good now.
